File: src/Core/ClashProxyStore.cs
using System.Text.Json;

namespace CodexSwitch.Core;

public sealed class ClashProxySettings
{
    public int HttpPort { get; set; } = 1990;
    public int SocksPort { get; set; } = 1991;
    public string Controller { get; set; } = "127.0.0.1:1993";
    public string LimiterListen { get; set; } = "127.0.0.1:1994";
    public string RelayGroup { get; set; } = "relay-group";
    public string TargetName { get; set; } = "target-socks5";
    public string RelayYaml { get; set; } = "";
    public string HomeServer { get; set; } = "";
    public string HomePort { get; set; } = "1080";
    public string HomeUsername { get; set; } = "";
    public string HomePassword { get; set; } = "";
    public int MaxConcurrent { get; set; } = 8;
    public int DialIntervalMs { get; set; } = 250;
    public int QueueWaitS { get; set; } = 8;
    public string MihomoPath { get; set; } = "";
    public string PythonPath { get; set; } = "";
}

public sealed class ClashProxyStore
{
    public string Home { get; }
    public string StoreDirectory => Path.Combine(Home, "codex-switch");
    public string SettingsPath => Path.Combine(StoreDirectory, "clash-proxy.json");
    public string WorkDirectory => Path.Combine(StoreDirectory, "clash");
    public string YamlPath => Path.Combine(WorkDirectory, "ai.yaml");
    public string LimiterJsonPath => Path.Combine(WorkDirectory, "socks-limiter.json");
    public string LimiterPythonPath => Path.Combine(WorkDirectory, "socks-limiter.py");
    public string MihomoLogPath => Path.Combine(WorkDirectory, "mihomo.log");
    public string LimiterLogPath => Path.Combine(WorkDirectory, "limiter.log");
    public string MihomoPidPath => Path.Combine(WorkDirectory, "mihomo.pid");
    public string LimiterPidPath => Path.Combine(WorkDirectory, "limiter.pid");

    public ClashProxyStore(string? home = null) => Home = Path.GetFullPath(home ??
        Environment.GetEnvironmentVariable("CODEX_HOME") ??
        Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".codex"));

    private static readonly JsonSerializerOptions Json = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true
    };

    public ClashProxySettings Load()
    {
        if (!File.Exists(SettingsPath)) return new ClashProxySettings();
        try
        {
            return JsonSerializer.Deserialize<ClashProxySettings>(File.ReadAllText(SettingsPath), Json) ?? new ClashProxySettings();
        }
        catch (JsonException)
        {
            throw new InvalidOperationException("双跳代理配置文件无效。");
        }
    }

    public void Save(ClashProxySettings settings)
    {
        System.IO.Directory.CreateDirectory(StoreDirectory);
        var json = JsonSerializer.Serialize(settings, Json);
        var temp = SettingsPath + ".tmp-" + Guid.NewGuid().ToString("N");
        File.WriteAllText(temp, json);
        File.Move(temp, SettingsPath, true);
        TryRestrict(SettingsPath);
    }

    public DialerProxyFiles Materialize(ClashProxySettings settings)
    {
        var files = DialerProxyBuilder.Build(ToInput(settings));
        Directory.CreateDirectory(WorkDirectory);
        Write(YamlPath, files.Yaml);
        Write(LimiterJsonPath, files.LimiterJson);
        Write(LimiterPythonPath, files.LimiterPython);
        TryRestrict(LimiterJsonPath);
        TryRestrict(LimiterPythonPath);
        return files;
    }

    public static DialerProxyInput ToInput(ClashProxySettings settings) => new(
        settings.RelayYaml,
        settings.HomeServer,
        settings.HomePort,
        settings.HomeUsername,
        settings.HomePassword,
        settings.RelayGroup,
        settings.TargetName,
        settings.HttpPort,
        settings.SocksPort,
        settings.Controller,
        settings.LimiterListen,
        settings.MaxConcurrent,
        1,
        settings.DialIntervalMs,
        settings.QueueWaitS);

    private static void Write(string path, string text)
    {
        var temp = path + ".tmp-" + Guid.NewGuid().ToString("N");
        File.WriteAllText(temp, text);
        File.Move(temp, path, true);
    }

    private static void TryRestrict(string path)
    {
        if (!OperatingSystem.IsWindows())
        {
            try { File.SetUnixFileMode(path, UnixFileMode.UserRead | UnixFileMode.UserWrite); } catch { }
        }
    }
}

File: src/macOS/Sources/CodexSwitchCore/ClashProxyStore.swift
import Foundation

public struct ClashProxySettings: Codable, Sendable, Equatable {
    public var httpPort = 1990
    public var socksPort = 1991
    public var controller = "127.0.0.1:1993"
    public var limiterListen = "127.0.0.1:1994"
    public var relayGroup = "relay-group"
    public var targetName = "target-socks5"
    public var relayYaml = ""
    public var homeServer = ""
    public var homePort = "1080"
    public var homeUsername = ""
    public var homePassword = ""
    public var maxConcurrent = 8
    public var dialIntervalMs = 250
    public var queueWaitS = 8
    public var mihomoPath = ""
    public var pythonPath = ""
    public init() {}
}

public actor ClashProxyStore {
    public nonisolated let home: URL
    public nonisolated var directory: URL { home.appendingPathComponent("codex-switch", isDirectory: true) }
    public nonisolated var settingsURL: URL { directory.appendingPathComponent("clash-proxy.json") }
    public nonisolated var workDirectory: URL { directory.appendingPathComponent("clash", isDirectory: true) }
    public nonisolated var yamlURL: URL { workDirectory.appendingPathComponent("ai.yaml") }
    public nonisolated var limiterJSONURL: URL { workDirectory.appendingPathComponent("socks-limiter.json") }
    public nonisolated var limiterPythonURL: URL { workDirectory.appendingPathComponent("socks-limiter.py") }
    public nonisolated var mihomoLogURL: URL { workDirectory.appendingPathComponent("mihomo.log") }
    public nonisolated var limiterLogURL: URL { workDirectory.appendingPathComponent("limiter.log") }
    public nonisolated var mihomoPidURL: URL { workDirectory.appendingPathComponent("mihomo.pid") }
    public nonisolated var limiterPidURL: URL { workDirectory.appendingPathComponent("limiter.pid") }
    private let fm = FileManager.default

    public init(home: URL? = nil) {
        self.home = home ?? ProcessInfo.processInfo.environment["CODEX_HOME"].map { URL(fileURLWithPath: $0, isDirectory: true) }
            ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".codex", isDirectory: true)
    }

    public func load() throws -> ClashProxySettings {
        guard fm.fileExists(atPath: settingsURL.path) else { return ClashProxySettings() }
        do { return try JSONDecoder().decode(ClashProxySettings.self, from: Data(contentsOf: settingsURL)) }
        catch { throw StoreError("双跳代理配置文件无效。") }
    }

    public func save(_ settings: ClashProxySettings) throws {
        try fm.createDirectory(at: directory, withIntermediateDirectories: true, attributes: [.posixPermissions: 0o700])
        let data = try JSONEncoder().encode(settings)
        try data.write(to: settingsURL, options: .atomic)
        try fm.setAttributes([.posixPermissions: 0o600], ofItemAtPath: settingsURL.path)
    }

    public func materialize(_ settings: ClashProxySettings) throws -> DialerProxyFiles {
        let files = try DialerProxyBuilder.build(input(settings))
        try fm.createDirectory(at: workDirectory, withIntermediateDirectories: true, attributes: [.posixPermissions: 0o700])
        try files.yaml.write(to: yamlURL, atomically: true, encoding: .utf8)
        try files.limiterJSON.write(to: limiterJSONURL, atomically: true, encoding: .utf8)
        try files.limiterPython.write(to: limiterPythonURL, atomically: true, encoding: .utf8)
        try fm.setAttributes([.posixPermissions: 0o600], ofItemAtPath: limiterJSONURL.path)
        return files
    }

    public func input(_ settings: ClashProxySettings) -> DialerProxyInput {
        DialerProxyInput(relayYaml: settings.relayYaml, homeServer: settings.homeServer, homePort: settings.homePort,
                         homeUsername: settings.homeUsername, homePassword: settings.homePassword,
                         relayGroup: settings.relayGroup, targetName: settings.targetName, httpPort: settings.httpPort,
                         socksPort: settings.socksPort, controller: settings.controller, limiterListen: settings.limiterListen,
                         maxConcurrent: settings.maxConcurrent, dialIntervalMs: settings.dialIntervalMs, queueWaitS: settings.queueWaitS)
    }
}
